model/infinispan/src/main/java/org/keycloak/infinispan/health/impl/ClusterHealthImpl.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2025 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.infinispan.health.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Objects;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.remoting.transport.Transport;
+import org.infinispan.remoting.transport.jgroups.JGroupsTransport;
+import org.infinispan.util.concurrent.BlockingManager;
+import org.jboss.logging.Logger;
+import org.keycloak.infinispan.health.ClusterHealth;
+import org.keycloak.jgroups.protocol.KEYCLOAK_JDBC_PING2;
+import org.keycloak.jgroups.protocol.KEYCLOAK_JDBC_PING2.HealthStatus;
+
+/**
+ * A {@link ClusterHealth} implementation that makes use of {@link KEYCLOAK_JDBC_PING2}.
+ * <p>
+ * Since each node is registered in the database, it is possible to detect if a partition is happening.
+ * <p>
+ * The method {@link KEYCLOAK_JDBC_PING2#isHealthy()} contains the algorithm description. If it returns
+ * {@link HealthStatus#ERROR}, the healthy state does not change and relies on the Quarkus/Agroal readiness health
+ * check. But, if {@link HealthStatus#NO_COORDINATOR} is returned, the state will be changed to unhealthy. The should be
+ * a temporary situation as at least one coordinator must be present in the database table.
+ *
+ * @see KEYCLOAK_JDBC_PING2#isHealthy()
+ */
+@Scope(Scopes.GLOBAL)
+public class ClusterHealthImpl implements ClusterHealth {
```

Comment:
As this is tightly coupled with JDBC_PING2, how about `JdbcPingClusterHealthImpl`?